Convert QuickStarts localstorage to user settings

sahil143 opened the pull request:
**Fixes**: https://issues.redhat.com/browse/ODC-5034
<!-- For e.g Fixes: https://issues.redhat.com/browse/ODC-XXX -->

**Analysis / Root cause**: move localstorage to user settings
<!-- Briefly describe analysis of US/Task or root cause of Defect -->

**Solution Description**: moved localStorage for quickstarts to userSettings
<!-- Describe your code changes in detail and explain the solution -->

**Screen shots / Gifs for design review**: No UI changes
<!-- If change affects UI in any way, tag @openshift/team-devconsole-ux and add screenshots/gifs  -->

**Unit test coverage report**: 
<!-- Attach test coverage report -->

**Test setup:**
<!-- If any setup required to test this PR, mention the details -->

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

invincibleJai commented:
/assign

abhinandan13jan commented on frontend/packages/console-app/src/components/quick-starts/catalog/QuickStartCatalog.tsx:
```diff
@@ -1,20 +1,21 @@
 import * as React from 'react';
-import { Dispatch } from 'redux';
-import { connect } from 'react-redux';
+// import { Dispatch } from 'redux';
```
Why are we keeping the comments?

abhinandan13jan commented on frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:
```diff
@@ -0,0 +1,116 @@
+export const QuickStartContext = createContext<QuickStartContextValues>({ activeQuickStartID: '' });
+
+const getInitialState = () =>
+  localStorage.getItem(QUICK_START_REDUX_STATE_LOCAL_STORAGE_KEY)
```
Should we still use data present over localStorage or is it better to start with scratch as in the future this could lead to anomaly caused by a user with malicious intent.

sahil143 replied:
this is a temporary code once the local storage is migrated over to user settings this will be removed.

abhinandan13jan commented on frontend/public/components/dashboard/dashboards-page/quick-start-badge.tsx:
```diff
@@ -1,18 +1,19 @@
 
 const HIDE_QUICK_START_BADGE_STORAGE_KEY = 'bridge/hide-quick-start-badge';
-
+const HIDE_QUICK_START_BADGE_CONFIG_MAP = 'console.quickStartBadge';
```
Just to confirm, Should HIDE_QUICK_START_BADGE_CONFIG_MAP and HIDE_QUICK_START_ADD_TILE_CONFIG_MAP_KEY be separate?

sahil143 replied:
What do you mean by separately hidden?

abhinandan13jan replied:
nvm

invincibleJai commented on frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:
```diff
@@ -0,0 +1,116 @@
+  return defaultQuickStartState;
+};
+
+const QUICK_START_KEY = 'console.quickstart';
+const ACTIVE_QUICK_START_ID_KEY = `${QUICK_START_KEY}.active`;
+const ALL_QUICK_START_STATE_KEY = `${QUICK_START_KEY}.allStates`;
```
nit: will it make sense to add these in const file for `quick-start`

sahil143 replied:
They are not being used in any other place. So it's fine to keep it here.

invincibleJai commented on frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:
```diff
@@ -0,0 +1,116 @@
+import { QUICK_START_REDUX_STATE_LOCAL_STORAGE_KEY } from '../../../redux/reducers/quick-start-reducer';
```
nit: sort imports and we can move this in a const file now, WDYT

invincibleJai commented on frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:
```diff
@@ -1,25 +1,13 @@
-  activeQuickStartState: QuickStartState;
-};
-
 type DispatchProps = {
```
can we remove `DispatchProps` and `MergeProps` as well now?

invincibleJai commented on frontend/packages/console-app/src/components/quick-starts/QuickStartDrawer.tsx:
```diff
@@ -27,14 +18,13 @@ type DispatchProps = {
-  activeQuickStartStatus,
-  onClose,
-}) => {
+const QuickStartDrawer: React.FC<QuickStartDrawerProps> = ({ children }) => {
```
looks like type definition `QuickStartDrawerProps` has extra props now here and even in other components which are no longer connected one

jerolimov commented:
/retest

invincibleJai commented:
locally frontend passed for me 😕

christianvogt commented:
Run i18n again .

sahil143 commented:
fixing unit tests

invincibleJai commented on frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:
```diff
@@ -1,58 +1,32 @@
+  } = React.useContext<QuickStartContextValues>(QuickStartContext);
+  const status = activeQuickStartState?.status as QuickStartStatus;
+  const taskNumber = activeQuickStartState?.taskNumber as number;
+  const allTaskStatuses = quickStart.spec.tasks.map(
```
nit: we have `tasks` destructured above , can use it and will tasks always be there or a list if not good to have check

invincibleJai commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/cluster-dashboard.tsx:
```diff
@@ -14,29 +14,28 @@ import { useK8sGet } from '../../../utils/k8s-get-hook';
   );
-  const [showQuickStartsCatalogCard, setShowQuickStartsCatalogCard] = React.useState<boolean>(
-    !isQuickStartsCatalogCardHidden,
+  // [TODO](sahil143): use sync capability here
```
will this be dependent on https://github.com/openshift/console/pull/7298 ?

sahil143 replied:
No, it doesn't have to as it only requires adding a third param which can be handled seperatly.

debsmita1 commented on frontend/packages/console-shared/src/components/quick-starts/__tests__/QuickStartsCatalogCard.spec.tsx:
```diff
@@ -49,8 +85,21 @@ describe('QuickStartsCatalogCard', () => {
     expect(emptyWrapper.find(QuickStartsCatalogCard).exists()).toBe(false);
   });
+
+  it('should not render is userSettings loaded is false', () => {
```
```suggestion
  it('should not render if userSettings loaded is false', () => {
```

debsmita1 commented on frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:
```diff
@@ -0,0 +1,116 @@
+
+export type QuickStartContextValues = {
+  activeQuickStartID: string;
+  allQuickStartStates?: any;
```
type ?

debsmita1 commented on frontend/packages/console-app/src/components/quick-starts/utils/quick-start-context.ts:
```diff
@@ -0,0 +1,116 @@
+  setQuickStartTaskNumber?: (quickStartId: string, taskNumber: number) => void;
+  setQuickStartTaskStatus?: (taskStatus: QuickStartTaskStatus) => void;
+  resetQuickStart?: (quickStartId: string, totalTasks: number) => void;
+  getQuickStartForId?: (id: string) => any;
```
type ?

jerolimov commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/cluster-dashboard.tsx:
```diff
@@ -45,7 +44,7 @@ export const ClusterDashboard: React.FC<{}> = () => {
                       quickStarts={quickStarts}
                       storageKey={HIDE_QUICK_START_DASHBOARD_TILE_STORAGE_KEY}
-                      onRemoveTile={onQuickStartsCatalogCardRemove}
+                      configMapKey={HIDE_QUICK_START_DASHBOARD_TILE_CONFIG_MAP_KEY}
```
```suggestion
                      userSettingsKey={HIDE_QUICK_START_DASHBOARD_TILE_USERSETTINGS_KEY}
```

jerolimov commented:
Tested this locally and it works fine. When you adopt Jai and Debsmitas, and maybe mine, I would add a lgtm.

@sahil143 @invincibleJai because the quick starts was never unmounted the localStorage key was not removed, right?

debsmita1 commented:
@sahil143  I came across a scenario in the cluster where there are no quick starts, because of which, in the cluster overview page there is a gap above the Activity card. So the condition `loaded && showQuickStartsCatalogCard` is not enough, we would also need to check if there are allowed quick starts available.

![cd](https://user-images.githubusercontent.com/22490998/100424185-3ae07580-30b3-11eb-8225-5e9509d8aad3.png)

sahil143 commented:
> @sahil143 @invincibleJai because the quick starts was never unmounted the localStorage key was not removed, right?

seems like it.

@jerolimov @debsmita1 @invincibleJai Addressed all review comments. Please review again.

invincibleJai commented on frontend/packages/console-app/src/components/quick-starts/catalog/QuickStartCatalog.tsx:
```diff
@@ -1,45 +1,24 @@
 import QuickStartTile from './QuickStartTile';
 
 import './QuickStartCatalog.scss';
+import { QuickStartContext, QuickStartContextValues } from '../utils/quick-start-context';
```
nit: sort import

jerolimov replied:
@sahil143 this one could still be improved

invincibleJai commented on frontend/packages/console-shared/src/components/quick-starts/QuickStartsCatalogCard.tsx:
```diff
@@ -17,54 +16,42 @@ import {
+  QuickStartContextValues,
+} from '@console/app/src/components/quick-starts/utils/quick-start-context';
 import './QuickStartsCatalogCard.scss';
+import { useUserSettingsCompatibility } from '../../hooks';
```
nit: sort imports

jerolimov replied:
fixed

invincibleJai commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/cluster-dashboard.tsx:
```diff
@@ -14,29 +14,28 @@ import { useK8sGet } from '../../../utils/k8s-get-hook';
 import { InfrastructureModel } from '../../../../models';
 import { K8sResourceKind } from '../../../../module/k8s';
 import { ClusterDashboardContext } from './context';
+import { useUserSettings } from '@console/shared';
```
nit: sort imports

jerolimov replied:
fixed

invincibleJai commented:
Verified the changes , works as expected cc @jerolimov

jerolimov commented:
Tested it locally, work fine. :+1: 

/lgtm

debsmita1 commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/cluster-dashboard.tsx:
```diff
@@ -19,24 +19,22 @@ const mainCards = [{ Card: StatusCard }, { Card: UtilizationCard }];
+    true,
   );
 
-  const onQuickStartsCatalogCardRemove = () => {
```
@sahil143 when the quick start card is removed, without this it leaves a gap above the Activity card as in the gird doesn't re-arrange itself properly

sahil143 replied:
Yeah, This should be fixed once the sync capability is available. userSettings will update the values once the quickStart card is removed.

sahil143 replied:
This is a one-line change which can go in with sync PR as well

invincibleJai commented:
/lgtm

christianvogt commented:
/approve
